### ISSUE TITLE
fairseq generated translations contain repeated sentences and commas

### ISSUE THREAD

**Issue by tianshuailu** [facebookresearch/fairseq#4637]
## ❓ Questions and Help

I tried to use fairseq-generate to see the performance of the model and the translations in the Hypothesis line contains many repeated lines, not sure if it is a problem with the model or the command I used which is shown below.

`fairseq-generate /srv/scratch3/ltian/data_trans/NMTAdapt \
  --path /srv/scratch3/ltian/train_nmt/checkpoint_10_28000pt.pt \
  --results-path /srv/scratch3/ltian/nmt_eval_result \
  --task translation_from_pretrained_bart \
  --gen-subset test \
  -t ml_XX -s en_XX \
  --bpe 'sentencepiece' --sentencepiece-model /srv/scratch3/ltian/sentence.bpe.model \
  --scoring sacrebleu \
  --batch-size 32 --langs ar_AR,cs_CZ,de_DE,en_XX,es_XX,et_EE,fi_FI,fr_XX,gu_IN,hi_IN,it_IT,ja_XX,kk_KZ,ko_KR,lt_LT,lv_LV,my_MM,ne_NP,nl_XX,ro_RO,ru_RU,si_LK,ml_XX,no_ML,no_HI`

some examples of the output

> S-1399	[' There',' is',' no',' students',' politics','.'][en_XX]
T-1399	[' ई',' विद्या','र्','\u200d','थि',' मु','न्','ने','ऱ','्','ऱ','त्','ति','नु',' राष्ट्रीय','मि','ल्ल','.']
H-1399	-219.1944580078125	'▁छात्र', '▁राजनीति', '▁में', '▁शामिल', '▁हैं', '.'] [hi_IN] ['▁इसमें', '▁कोई', '▁भी', '▁नहीं', '▁है', '।'] [hi_IN] ['▁उन्होंने', '▁कहा', '▁है', '।'] [hi_IN] ['▁इसमें', '▁कोई', '▁भी', '▁नहीं', '▁है', '।'] [hi_IN] ["▁'", [hi_IN] ["▁'", [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ["▁'", [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ["▁'", [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्र', '▁हैं', '।'] [hi_IN] ['▁छात्र', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁हैं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁है', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '▁है', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '.'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁है', '।'] [hi_IN] ['▁छात्रों', '▁की', '▁राजनीति', '▁में', '▁भी', '▁नहीं', '।']

> S-1753	[' This',' relationship',' then',' became',' warm','.'][en_XX]
T-1753	[' पि','न्न','ी','ट्',' ई',' ','बन्ध','ं',' कू','ट','ु','त','ल्','\u200d',' ','दृ','ड','मा','य','ि','.']
H-1753	-278.149658203125	['▁तब', '▁यह', '▁संबंध', '▁गर्म', '▁हो', '▁गया', '।'] [hi_IN] [hi_IN] ['▁फिर', '।'] [hi_IN] ['▁इस', '▁बीच', '▁में', '▁यह', '▁रिश्', '\u200d', 'या', '▁है', '।'] [hi_IN] ['▁फिर', '।'] [hi_IN] ['▁फिर', '।'] [hi_IN] ['▁इस', '▁बीच', '▁में', '▁यह', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁और', '▁यह', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁तब', '।'] [hi_IN] ['▁तब', '।'] [hi_IN] ['▁तब', '।'] [hi_IN] ['▁तब', '।'] [hi_IN] ['▁तब', '।'] [hi_IN] ['▁इस', '▁बीच', '▁में', '▁यह', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁तब', '▁से', '▁हुआ', '।'] [hi_IN] ['▁तब', '▁से', '▁हुआ', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁से', '▁हुआ', '।'] [hi_IN] ['▁इस', '▁तरह', '▁से', '▁हुआ', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁से', '▁हुआ', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁ये', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁से', '▁हुआ', '।'] [hi_IN] ['▁इस', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁इस', '▁संबंध', '▁गर्म', '▁हो', '▁गया', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁ये', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁इसके', '▁बाद', '▁ये', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁इसके', '▁लिए', '▁गए', '▁थे', '।'] [hi_IN] ['▁इसके', '▁लिए', '▁गए', '▁थे', '।'] [hi_IN] ['▁इस', '▁रिश्', '\u200d', 'या', '▁गया', '।'] [hi_IN] ['▁इस', '▁रिश्', '\u200d', 'या', '▁गया', '।']

And clearly, the translation is way too long. If it helps, the model is trained on hindi-en parallel data and malayalam-en back translation data, the target in the example is transliterated malayalam with Devanagari script

#### What have you tried?
I tried to set --lenpen parameter as smaller than 0 hoping that it will reduce the length but it doesn't work
#### What's your environment?

 - fairseq Version v0.10.2:
 - PyTorch Version (e.g., 1.0)
 - OS (e.g., Linux):
 - How you installed fairseq (`pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by tianshuailu** [comment]
And also some of the translations contain many commas, I think it's because the back translation data, which we are supposed to use for training the model, contain many sentences like these, but any idea how to prevent model from generating sentences like these?

> S-623	[' Hospital',' authorities',' informed',' the',' police','.'][en_XX]
T-623	[' आ','शु','पत्र','ि',' अधिकृत','र्','\u200d',' वि','वर','ं',' पोल','ी','सि','न','ॆ',' अ','ऱ','िय','ि','च्च','ु','.']
H-623	-111.30795288085938	'▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", '▁""', "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'", "'",

> S-1197	[' The',' Bible','s',' advice',' really',' works'][en_XX]
T-1197	[' बै','बि','ळ','ि','न्','\u200d',<<unk>>' बुद्धि','यु','प','देश','ं',' त','िक','च्च','ु','ं',' प्रा','योग','िक','मा','ण्']
H-1197	-258.78936767578125	['▁“', '▁”', '“', '“', '▁”', '“', '“', '▁”', '“', '“', '▁”', '“', '“', '▁”', '“', '“', '▁”', '“', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '“', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '“', '▁”', '▁”', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '“', '“', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '▁”', '▁”', '“', '“', '“', '“', '“', '▁”', '“', '“', '“', '“', '“', '“', '“', '“', '“', '▁”', '▁”', '▁”', '“', '“', '“', '▁”', '“', '“', '“', '“', '“', '“', '“', '“',

**Comment by erip** [comment]
How did you preprocess your data? It seems like neither your source nor target are subword encoded and probably your model just didn't learn anything useful.

**Comment by gmryu** [comment]
I am not familiar with multilingual model. (Yours is multilingual right?)

You may have to provide `--decoder-langtok --encoder-langtok src` .

(also it would be nice of you to include "multilingual" or something close into the issue title if it is so.)

**Comment by tangyuq** [comment]
how do you train your model? or which mbart models are finetuned from? 

If you are finetuning from mbart50, you might want to try:  -- task translation_multi_simple_epoch 
`fairseq-generate /srv/scratch3/ltian/data_trans/NMTAdapt \ --path /srv/scratch3/ltian/train_nmt/checkpoint_10_28000pt.pt \ --results-path /srv/scratch3/ltian/nmt_eval_result \ --task **translation_multi_simple_epoch** \ --gen-subset test \ -t ml_XX -s en_XX \ --bpe 'sentencepiece' --sentencepiece-model /srv/scratch3/ltian/sentence.bpe.model \ --scoring sacrebleu \ --batch-size 32 --langs ar_AR,cs_CZ,de_DE,en_XX,es_XX,et_EE,fi_FI,fr_XX,gu_IN,hi_IN,it_IT,ja_XX,kk_KZ,ko_KR,lt_LT,lv_LV,my_MM,ne_NP,nl_XX,ro_RO,ru_RU,si_LK,ml_XX,no_ML,no_HI`

If it is from mbart50, please check: https://github.com/facebookresearch/fairseq/tree/main/examples/multilingual

**Comment by tianshuailu** [comment]
> How did you preprocess your data? It seems like neither your source nor target are subword encoded and probably your model just didn't learn anything useful.

Here is the command for preprocessing the data. 

> DATA=/srv/scratch3/ltian/data_trans/bt
TRAIN=train 
VALID=valid 
SRC=en_XX 
TGT=ml_XX 
DICT=/srv/scratch3/ltian/dict.txt 
DEST=/srv/scratch3/ltian/data_trans 
NAME=NMTAdapt
fairseq-preprocess \
   --source-lang ${SRC} \
   --target-lang ${TGT} \
   --trainpref ${DATA}/${TRAIN}.spm \
   --validpref ${DATA}/${VALID}.spm \
   --destdir ${DEST}/${NAME} \
   --thresholdtgt 0 \
   --thresholdsrc 0 \
   --srcdict ${DICT} \
   --tgtdict ${DICT} \
   --workers 70

Before running these commands, we also did bpe to the sentences with sentencepiece. which looks like this

> ['▁The', '▁resident', 's', '▁said', '▁several', '▁houses', '▁were', '▁', 'burg', 'led', '▁in', '▁the', '▁area', '▁recently', '.']
['▁After', '▁that', ',', '▁the', '▁couple', '▁organised', '▁two', '▁grand', '▁re', 'ception', 's', '▁in', '▁Delhi', '▁and', '▁Mumbai', '.']
['▁नेता', 'वे', ',', '▁अ', 'त', 'त्र', '▁मार', 'कम', 'ॊ', 'न्', 'नु', 'म', 'ल्ल', '.']
['▁समय', 'बन्ध', 'ित', 'मा', 'य', 'ि', '▁पू', 'ऱ', '्', 'त्त', 'ी', 'कर', 'िक', '्', 'कु', 'ं', '.']
['▁संभव', 'स्थ', 'ल', 'ं', '▁मु', 'ऴ', 'ु', 'व', 'न्', '▁पोल', 'ी', 'स्', '▁निरीक्षण', 'त्त', 'िला', 'ण्', '.']

The process in this link is basically what we did. [https://github.com/wjko2/NMT-Adapt](url)
And also, our task is to train the translation model for low recourse languages, in our case it's Malayalam, so during the training we didn't use ml-en parallel data.
In short, we first pretrain mbart model with Hindi-English parallel data, and then use the pre-trained model to generate ml-en back translation data, and finally use hi-en, ml-en data to train the pre-trained model.

**Comment by tianshuailu** [comment]
> how do you train your model? or which mbart models are finetuned from?
> 
> If you are finetuning from mbart50, you might want to try: -- task translation_multi_simple_epoch `fairseq-generate /srv/scratch3/ltian/data_trans/NMTAdapt \ --path /srv/scratch3/ltian/train_nmt/checkpoint_10_28000pt.pt \ --results-path /srv/scratch3/ltian/nmt_eval_result \ --task **translation_multi_simple_epoch** \ --gen-subset test \ -t ml_XX -s en_XX \ --bpe 'sentencepiece' --sentencepiece-model /srv/scratch3/ltian/sentence.bpe.model \ --scoring sacrebleu \ --batch-size 32 --langs ar_AR,cs_CZ,de_DE,en_XX,es_XX,et_EE,fi_FI,fr_XX,gu_IN,hi_IN,it_IT,ja_XX,kk_KZ,ko_KR,lt_LT,lv_LV,my_MM,ne_NP,nl_XX,ro_RO,ru_RU,si_LK,ml_XX,no_ML,no_HI`
> 
> If it is from mbart50, please check: https://github.com/facebookresearch/fairseq/tree/main/examples/multilingual

Thanks for your suggestion, I'll definitely have a try.
I pretrained from mbart.cc25 model [https://github.com/facebookresearch/fairseq/tree/main/examples/mbart](url)
And the training process is from this link and also described in my last reply to erip. [https://github.com/wjko2/NMT-Adapt](url)

**Comment by tianshuailu** [comment]
> I am not familiar with multilingual model. (Yours is multilingual right?)
> 
> You may have to provide `--decoder-langtok --encoder-langtok src` .
> 
> (also it would be nice of you to include "multilingual" or something close into the issue title if it is so.)

We provided these parameters `--decoder-langtok --encoder-langtok src` in the training process, do you think we should also include it in the fairseq-generate command?
`fairseq-train  /srv/scratch3/ltian/data_trans/NMTAdapt   --encoder-normalize-before --decoder-normalize-before   --arch mbart_large --layernorm-embedding   \
--task translation_multi_simple_epoch    --criterion label_smoothed_cross_entropy --label-smoothing 0.2   --optimizer adam --adam-eps 1e-06 --adam-betas '(0.9, 0.98)' \
--lr-scheduler polynomial_decay --lr 3e-05 --min-lr -1 --warmup-updates 50 --total-num-update 80000   --dropout 0.3 --attention-dropout 0.1 --weight-decay 0.0 \
--max-tokens 512 --update-freq 8   --save-interval 1 --save-interval-updates 7000 --keep-interval-updates 10 --no-epoch-checkpoints   --seed 222 --log-format simple \
--log-interval 2  --save-dir /srv/scratch3/ltian/train_nmt \
--restore-file /srv/scratch3/ltian/pretrained_model.pt  --reset-optimizer --reset-meters --reset-dataloader --reset-lr-scheduler   \
--ddp-backend no_c10d  --max-epoch 128  --skip-invalid-size-inputs-valid-test --fp16 \
--langs ar_AR,cs_CZ,de_DE,en_XX,es_XX,et_EE,fi_FI,fr_XX,gu_IN,hi_IN,it_IT,ja_XX,kk_KZ,ko_KR,lt_LT,lv_LV,my_MM,ne_NP,nl_XX,ro_RO,ru_RU,si_LK,ml_XX,no_ML,no_HI \
--lang-pairs en_XX-hi_IN,en_XX-ml_XX,no_HI-hi_IN,no_ML-ml_XX  --virtual-epoch-size 1388932 --virtual-data-size 138893200 \
--sampling-weights "{'main:en_XX-hi_IN': 1,'main:en_XX-ml_XX': 1,'main:no_HI-hi_IN': 1,'main:no_ML-ml_XX':1}" --keep-inference-langtok --encoder-langtok src \
--decoder-langtok --lang-tok-style mbart --checkpoint-suffix pt`

And yeah, it is a multilingual task, thanks for the reminder, will include the tag next time.

**Comment by erip** [comment]
Does your spm encoded data actually contain a list like that per row? What does `head -n2 /srv/scratch3/ltian/data_trans/bt/train.spm.en_XX` show? These files should be whitespace-delimited spm encoded data like:

```
▁The ▁resident s ▁said ▁several ▁houses ▁were ▁ burg led ▁in ▁the ▁area ▁recently .
▁After ▁that , ▁the ▁couple ▁organised ▁two ▁grand ▁re ception s ▁in ▁Delhi ▁and ▁Mumbai .
...
```

**Comment by tianshuailu** [comment]
> Does your spm encoded data actually contain a list like that per row? What does `head -n2 /srv/scratch3/ltian/data_trans/bt/train.spm.en_XX` show? These files should be whitespace-delimited spm encoded data like:
> 
> ```
> ▁The ▁resident s ▁said ▁several ▁houses ▁were ▁ burg led ▁in ▁the ▁area ▁recently .
> ▁After ▁that , ▁the ▁couple ▁organised ▁two ▁grand ▁re ception s ▁in ▁Delhi ▁and ▁Mumbai .
> ...
> ```

Hi erip, thanks for pointing that out, the data does contain a list like that, this is what the head command show
![image](https://user-images.githubusercontent.com/22446406/183736611-22f42559-8ad9-47d4-8720-a43f67c11104.png)
I didn't realise that could be a problem, I thought that is how the data should look like after the bpe process. Do you might know a reliable way to remove all the brackets, quotes and commas?

**Comment by erip** [comment]
That doesn't look like valid sentencepiece output. You might want to restart... You can do something like:

```sh
cat train.en | spm_encode --model /path/to/spm.model > /path/to/train.spm.en_XX
```

for each file.

**Comment by tianshuailu** [comment]
> That doesn't look like valid sentencepiece output. You might want to restart... You can do something like:
> 
> ```shell
> cat train.en | spm_encode --model /path/to/spm.model > /path/to/train.spm.en_XX
> ```
> 
> for each file.

Thank you so much for finding out the problem and the command, it works for me.
